src/pruna/engine/utils.py

```diff
@@ -200,3 +201,70 @@ def set_to_train(model: Any) -> None:
         # Since after compression most of the models are inference only, the iteration could lead to unexpected behavior. # noqa: E501
         # This should be investigated in the future.
         pruna_logger.warning("Model does not support training mode.")
+
+
+class ModelContext:
+    """
+    Context manager for handling the model.
+
+    Parameters
+    ----------
+    model : ModelMixin
+        The model to handle. Can be a transformer model, UNet, or other ModelMixin.
+    """
+
+    def __init__(self, model: "ModelMixin") -> None:
+        """
+        Context manager for handling the model.
+
+        Parameters
+        ----------
+        model : ModelMixin
+            The model to handle. Can be a transformer model, UNet, or other pipeline.
+        """
+        self.pipeline = model
+
+    def __enter__(self) -> tuple[ModelMixin, Any, str | None]:
+        """
+        Enter the context manager.
+
+        Returns
+        -------
+        ModelMixin
+            The working model.
+        Any
+            The denoiser type.
+        str | None
+            The denoiser type.
+        """
+        if hasattr(self.pipeline, "transformer"):
+            self.working_model = self.pipeline.transformer
+            self.denoiser_type = "transformer"
+        elif hasattr(self.pipeline, "unet"):
+            self.working_model = self.pipeline.unet
+            self.denoiser_type = "unet"
+        else:
+            self.working_model = self.pipeline
+            self.denoiser_type = None  # type: ignore [assignment]
+        return self.pipeline, self.working_model, self.denoiser_type
+
+    def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
+        """
+        Exit the context manager.
+
+        Parameters
+        ----------
+        exc_type : Exception
+            The exception type.
+        exc_value : Exception
+            The exception value.
+        traceback : Exception
+            The traceback.
+        """
+        if hasattr(self.pipeline, "transformer"):
+            self.pipeline.transformer = self.pipeline.working_model
```

Comment:
I think this should be 
`self.pipeline.transformer = self.working_model`

Reply:
Maybe I am missing something especially since I am not super familiar with how these context classes work but in the `__enter__` method working model is set as an attribute of the ModelContext and not as an attribute of the pipeline. So I think `__exit__` should be 
```suggestion
        if hasattr(self.pipeline, "transformer"):
            self.pipeline.transformer = self.working_model
            elif hasattr(self.pipeline, "unet"):
            self.pipeline.unet = self.working_model
        else:
            self.pipeline = self.working_model
        del self.working_model
```
Alternatively you can set `working_model` as an attribute of `self.pipeline` in the `__enter__` method.

Reply:
Yeah I see what you mean ;) tbh I am not familiar with context manager, and during my test I found that modifying the working_model breaks the attribute link in the context manager. I tried to explain this in [this comment](https://github.com/PrunaAI/pruna/blob/9e5f43bf5fa992fd8f345c321f9b0d7d98395658/src/pruna/algorithms/factorizing/qkv_factorizing.py#L98) :(
Ideally, I agree with you I would like to just modify the working_model, and do self.pipeline.transformer = self.working_model in the `__exit__`.
Instead I have to define the working_model (once it is smashed) as an attribute of the pipeline like [this](https://github.com/PrunaAI/pruna/blob/9e5f43bf5fa992fd8f345c321f9b0d7d98395658/src/pruna/algorithms/factorizing/qkv_factorizing.py#L100), and then do self.pipeline.transformer = self.pipeline.working_model  in the `__exit__`

Reply:
I completely missed your request for my review—sorry about that. Overall, this looks fine to me. However, like you, I’m not entirely happy with the current implementation, since it requires you to set pipeline.working_model inside the context block, and will fail otherwise. I’m not sure that’s always desirable. Moreover, if the working model is itself an nn.Module, attaching one nn.Module as an attribute of another could introduce side effects.

I also wonder whether you really need the __exit__ method at all, since qkv-fusing is an in-place operation of the transformer and may not require any overwrite. But if you’d prefer to keep the version as-is, that works for me too.